### PR TITLE
Simulation of Joins and Uplinks

### LINKED ISSUE #361: Uplink message simulation
**Summary:** 

Our v1/v2 ttnctl used to have commands for simulating uplink messages. Let's build something similar for v3

**Why do we need this?**

This can be extremely useful for testing and debugging end-to-end flows, because you'd no longer need to connect a gateway and device.

**What is already there? What do you see now?**

We don't have this yet in v3

**What is missing? What do you want to see?**

Commands in the CLI to send simulated (re)join and uplink messages.

**How do you propose to implement this?**

- Build the LoRaWAN message from CLI flags and defaults
- Build the uplink metadata from CLI flags and defaults
- Connect to the Gateway Server's LinkGateway RPC (identifying as one of the user's gateways)
- Send the message with the metadata
- Wait for downlink(s) and print to terminal

**What can you do yourself and what do you need help with?**

I can implement an MVP. @johanstokking had some ideas for extending this in the future, so that we can also simulate messages being received on multiple gateways.

### PR DESCRIPTION
**Pull request by htdvisser**
**Summary:**

This PR closes #361 by adding (experimental) CLI commands for join and uplink simulation.

**Notes for Reviewers:**

This will (obviously) be extended in the future

### CODE REVIEW

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+
+	logger.Info("Sent uplink")
+
+	expect, _ := cmd.Flags().GetInt("downlinks")
+	for i := 0; i < expect; i++ {
+		down, err := link.Recv()
+		if err != nil {
+			return err
+		}
+		timestampDifference := uint32(uint64(down.GetDownlinkMessage().GetScheduled().GetTimestamp()) + 1<<32 - uint64(upMsg.GetRxMetadata()[0].GetTimestamp()))
```
Comment (KrishnaIyer) [inline cmd/ttn-lw-cli/commands/simulate.go]: What's the offset `1<<32` meant for? Also that doesn't seem right. `1<<32` exceeds `uint32` and will offset the timestamp by a 100 years or so. Is this intended?

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+		logger.Infof("Received downlink after %s for %s after uplink", time.Since(sendTime), time.Duration(timestampDifference)*1000)
```
Comment (KrishnaIyer) [inline cmd/ttn-lw-cli/commands/simulate.go]: I think this message can be slightly improved as below:
`Received downlink after %s and %s relative to the uplink`

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+	sendTime := time.Now()
+	if err = link.Send(&ttnpb.GatewayUp{UplinkMessages: []*ttnpb.UplinkMessage{upMsg}}); err != nil {
+		return err
+	}
+
+	logger.Info("Sent uplink")
+
+	expect, _ := cmd.Flags().GetInt("downlinks")
+	for i := 0; i < expect; i++ {
+		down, err := link.Recv()
```
Comment (KrishnaIyer) [inline cmd/ttn-lw-cli/commands/simulate.go]: I guess concurrency is for the future or is it not needed at all? The caller can optionally set the Uplink timestamp so Downlinks can arrive out of sequence right?

File: cmd/ttn-lw-cli/commands/simulate.go [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+	"go.thethings.network/lorawan-stack/pkg/band"
+	"go.thethings.network/lorawan-stack/pkg/crypto"
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+type simulateUplinkParams struct {
```
[the remark this answers is not in the document]
Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/simulate.go]: The `params` suffix is for consistency with the structs below. I guess we could rename it to `simulateMetadataParams` but to be honest I don't think it's worth the time

File: cmd/ttn-lw-cli/commands/simulate.go (hunk 1 above)
Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/simulate.go]: This is intended. The `1<<32` is added to the downlink timestamp (as `uint64`) after which the uplink timestamp is subtracted. If there was no overflow in downlink timestamp there is indeed `1<<32` too much, but the conversion back to `uint32` simply ignores that extra `1<<32`. If there _was_ overflow in the downlink timestamp, the result is the correct timestamp difference.

File: cmd/ttn-lw-cli/commands/simulate.go (hunk 3 above)
Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/simulate.go]: With the current simulation we don't expect any downlinks before we send the uplink. Also, we don't (yet) support sending multiple uplinks, that's out of scope for these simulation commands.

File: cmd/ttn-lw-cli/commands/simulate.go (hunk 1 above)
Comment (KrishnaIyer) [reply cmd/ttn-lw-cli/commands/simulate.go]: Ok but why would there be an overflow though? 
Unix time will not cross uint32 until 2048. So overflows mean that the DL timestamp is invalid right?

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/simulate.go]: This has nothing to do with Unix time. This is the concentrator timestamp.

Comment (KrishnaIyer) [reply cmd/ttn-lw-cli/commands/simulate.go]: Alright. LGTM

**Comment by Recodify** [comment]
Is this feature now available in the current version of ttn-lw-cli that is installed by snap (3.10.0-rc2)? I can't find any reference to it?

**Comment by neoaggelos** [comment]
Hi @Recodify,

Yes, this is available as an expiremental (and hidden command). You can try running `ttn-lw-cli simulate join-request --help` or `ttn-lw-cli simulate uplink --help` for more details. However, note that these commands are directed more towards internal debugging and testing for the Network Server.

If you want to test application-level features (e.g. payload formatters, integrations) instead, you might find that the uplink message simulation feature that was added to the console in 3.10 (Devices > [Device] > Messaging > Uplink > Simulate Uplink) would be of more use.

**EDIT**: See https://github.com/TheThingsNetwork/lorawan-stack/pull/3266

**Pull request merged**